pkg/aws/plugin.go

```diff
@@ -0,0 +1,258 @@
+		return nil, fmt.Errorf("resource description should not contain subnet ID")
+	}
+	if runInstancesInput.PrivateIpAddress != nil || runInstancesInput.EnablePrimaryIpv6 != nil ||
+		runInstancesInput.Ipv6AddressCount != nil || len(runInstancesInput.Ipv6Addresses) > 0 {
+		return nil, fmt.Errorf("resource description should not contain IP address information")
+	}
+
+	// Get region
+	availabilityZone := *runInstancesInput.Placement.AvailabilityZone
+	region := availabilityZone[:len(availabilityZone)-1]
+
+	// Load config and setup clients
+	cfg, err := config.LoadDefaultConfig(ctx, config.WithRegion(region))
+	if err != nil {
+		return nil, fmt.Errorf("unable to load config: %w", err)
+	}
+	ec2Client := awsClients.getOrCreateEc2Client(cfg)
+
+	// Get existing VPC
+	var vpc *types.Vpc
+	var subnet *types.Subnet
+	vpcName := getVpcName(req.Deployment.Namespace, region)
+	describeVpcsOutput, err := ec2Client.DescribeVpcs(ctx, &ec2.DescribeVpcsInput{
+		Filters: getDescribeFilter(req.Deployment.Namespace, vpcName),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to get VPCs: %w", err)
+	}
+	if len(describeVpcsOutput.Vpcs) <= 1 {
+		if len(describeVpcsOutput.Vpcs) == 1 {
+			vpc = &describeVpcsOutput.Vpcs[0]
+		} else {
+			// Find unused address spaces from orchestrator
+			orchestratorConn, err := grpc.NewClient(s.orchestratorServerAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+			if err != nil {
+				return nil, fmt.Errorf("unable to establish connection with orchestrator: %w", err)
+			}
+			defer orchestratorConn.Close()
+			orchestratorClient := paragliderpb.NewControllerClient(orchestratorConn)
+			findUnusedAddressSpacesReq := &paragliderpb.FindUnusedAddressSpacesRequest{}
+			findUnusedAddressSpacesResp, err := orchestratorClient.FindUnusedAddressSpaces(ctx, findUnusedAddressSpacesReq)
+			if err != nil {
+				return nil, fmt.Errorf("unable to find unused address spaces from orchestrator: %w", err)
+			}
+			vpcCidrBlock := findUnusedAddressSpacesResp.AddressSpaces[0]
+
+			// Create VPC
+			createVpcInput := &ec2.CreateVpcInput{
+				CidrBlock:         aws.String(vpcCidrBlock),
+				TagSpecifications: getCreateTagSpecifications(req.Deployment.Namespace, vpcName, types.ResourceTypeVpc),
+			}
+			createVpcOutput, err := ec2Client.CreateVpc(ctx, createVpcInput)
+			if err != nil {
+				return nil, fmt.Errorf("unable to create VPC: %w", err)
+			}
+			vpc = createVpcOutput.Vpc
+
+			// Wait until default security group is available
+			securityGroupExistsWaiter := ec2.NewSecurityGroupExistsWaiter(ec2Client)
+			err = securityGroupExistsWaiter.Wait(ctx, &ec2.DescribeSecurityGroupsInput{
+				Filters: []types.Filter{
+					{Name: aws.String("vpc-id"), Values: []string{*createVpcOutput.Vpc.VpcId}},
+				},
+			}, 10*time.Second)
+			if err != nil {
+				return nil, fmt.Errorf("unable to wait for default security group: %w", err)
+			}
+
+			// Remove default inbound and outbound rules from default security group
```

Comment:
Would it be easier to just put a deny-all rule at a higher priority than these defaults? That's what we do in Azure.

Reply:
I don't think AWS has deny rules for security groups. From https://docs.aws.amazon.com/vpc/latest/userguide/security-group-rules.html#security-group-rule-characteristics

> You can specify allow rules, but not deny rules.

There are network ACLs (which have both allow and deny) but I believe they apply over an entire subnet? Let me know if I'm wrong about that though cus I'm not 100% sure.